server/protocols/jmap/src/main/java/org/apache/james/jmap/mailet/filter/MailMatcher.java

```diff
@@ -0,0 +1,245 @@
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableMap;
+
+public interface MailMatcher {
+
+    interface HeaderExtractor extends ThrowingFunction<Mail, Stream<String>> {}
+
+    class HeaderMatcher implements MailMatcher {
+
+        private final Logger logger = LoggerFactory.getLogger(HeaderMatcher.class);
+
+        private final ContentMatcher contentMatcher;
+        private final String ruleValue;
+        private final HeaderExtractor headerExtractor;
+
+        HeaderMatcher(Optional<ContentMatcher> contentMatcherOptional, String ruleValue,
+                      Optional<HeaderExtractor> headerExtractorOptional) {
+            Preconditions.checkArgument(contentMatcherOptional.isPresent());
+            Preconditions.checkArgument(headerExtractorOptional.isPresent());
+
+            this.contentMatcher = contentMatcherOptional.get();
+            this.ruleValue = ruleValue;
+            this.headerExtractor = headerExtractorOptional.get();
+        }
+
+        @Override
+        public boolean match(Mail mail) {
+            try {
+                Stream<String> headerLines = headerExtractor.apply(mail);
+                return contentMatcher.match(headerLines, ruleValue);
+            } catch (Exception e) {
+                logger.error("error while extracting mail header", e);
+                return false;
+            }
+        }
+    }
+
+    interface ContentMatcher {
+
+        class AddressHeader {
+
+            static class Builder {
+                private final Logger logger = LoggerFactory.getLogger(Builder.class);
+
+                private InternetAddress internetAddress;
+
+                private Builder() {
+                }
+
+                private Builder internetAddress(String addressAsString) {
+                    try {
+                        this.internetAddress = new InternetAddress(addressAsString);
+                    } catch (AddressException e) {
+                        logger.error("error while parsing address " + addressAsString, e);
+                    }
+
+                    return this;
+                }
+
+                public AddressHeader build() {
+                    Preconditions.checkNotNull(internetAddress);
+
+                    return new AddressHeader(internetAddress.getPersonal(), internetAddress.getAddress());
+                }
+            }
+
+            public static Builder builder() {
+                return new Builder();
+            }
+
+            private final String personal;
+            private final String address;
+
+            private AddressHeader(String personal, String address) {
+                this.personal = personal;
+                this.address = address;
+            }
+
+            public String getPersonal() {
+                return personal;
+            }
+
+            public String getAddress() {
+                return address;
+            }
+
+            public String asString() {
```

Comment:
Why not keeping the original string passed to `internetAddress` method of the builder?

This could would then no more have knowledge of address format, which is great!